server/src/main/java/io/spine/server/integration/EventsExchange.java

```diff
@@ -0,0 +1,135 @@
+ *
+ * <p>Receives the domain events from other Bounded Contexts and posts them to the domestic
+ * bus treating them as {@code external}.
+ */
+class EventsExchange extends AbstractExchange {
+
+    private final BusAdapter bus;
+
+    /**
+     * Creates a new exchange which uses the passed link.
+     */
+    EventsExchange(TransportLink link, BusAdapter bus) {
+        super(link);
+        this.bus = bus;
+    }
+
+    /**
+     * Publishes the given event via this exchange.
+     *
+     * <p>Does nothing for events, which by chance, were passed into this method, being originated
+     * from other Bounded Contexts.
+     *
+     * @param event
+     *         event to publish
+     */
+    @Internal
+    void publish(EventEnvelope event) {
+        ChannelId channelId = toChannelId(event.messageClass());
+        boolean wantedByOthers = !subscriptionChannels().contains(channelId);
+        if (wantedByOthers) {
+            Event outerObject = event.outerObject();
+            ExternalMessage msg = ExternalMessages.of(outerObject, context());
+            Publisher publisher = publisher(channelId);
+            publisher.publish(AnyPacker.pack(event.id()), msg);
+        }
+    }
+
+    /**
+     * Registers a local dispatcher which would like to receive the {@code external} events
+     * through this exchange.
+     *
+     * @param dispatcher
+     *         the dispatcher to register
+     */
+    void register(EventDispatcher dispatcher) {
+        Iterable<EventClass> receivedTypes = dispatcher.externalEventClasses();
+        for (EventClass cls : receivedTypes) {
+            ChannelId channelId = toChannelId(cls);
+            Subscriber subscriber = subscriber(channelId);
+            IncomingEventObserver observer = observerFor(cls);
+            subscriber.addObserver(observer);
+        }
+    }
+
+    /**
+     * Unregisters a local dispatcher in case it should no longer receive its
+     * {@code external} events via this exchange.
+     *
+     * @param dispatcher
+     *         the dispatcher to unregister
+     */
+    public void unregister(EventDispatcher dispatcher) {
+        Iterable<EventClass> externalEvents = dispatcher.externalEventClasses();
+        for (EventClass cls : externalEvents) {
+            ChannelId channelId = toChannelId(cls);
+            Subscriber subscriber = subscriber(channelId);
+            IncomingEventObserver observer = observerFor(cls);
+            subscriber.removeObserver(observer);
+        }
+    }
+
+    /**
+     * Creates a new observer for a particular type of events.
+     */
+    private IncomingEventObserver observerFor(EventClass eventType) {
+        IncomingEventObserver observer = new IncomingEventObserver(context(), eventType, bus);
+        return observer;
+    }
+
+    /**
+     * Creates an ID of the channel that will transmit the events of the given class.
+     */
+    private static ChannelId toChannelId(EventClass cls) {
+        checkNotNull(cls);
```

Comment:
This null check looks as excessive, assuming we're doing it in `private static` method, and the next line calls a method of the passed object, anyway.